lib/Net/Async/Blockchain/BTC.pm

```diff
@@ -187,54 +178,62 @@ you received.
 =cut
 
 async sub recursive_search {
-    my ($self) = @_;
+    my ($self, $block_number) = @_;
 
-    return undef unless $self->base_block_number;
+    # that is nothing to do
+    return undef unless $block_number;
 
     my $current_block = await $self->rpc_client->get_last_block();
-    die 'unable to get the last block from the node.' unless $current_block;
+    # no matter if it is an error or not we can't proceed without the last block
+    return $block_number unless $current_block;
 
-    my $block_number_counter = $self->base_block_number;
+    my $block_number_counter = $block_number;
     while ($current_block >= $block_number_counter) {
         my $block_hash = await $self->rpc_client->get_block_hash($block_number_counter);
-        die "Failed to get the block hash for block number: $block_number_counter" unless $block_hash;
+        return $block_number_counter unless $block_hash;
+
+        # 2 here means the full verbosity since we want to get the raw transactions
+        my $block_response = await $self->rpc_client->get_block($block_hash, 2);
+        return $block_number_counter unless $block_response;
 
-        my $block_number = await $self->hashblock($block_hash);
-        $self->emit_block($block_number);
+        push($self->{pending_transactions}->@*, $block_response->{tx}->@*);
+        $self->emit_block($block_number_counter);
```

Comment:
So you are updating the block number only here? this mean between the two runs could have many many blocks and process transactions you already have received :confused: 
I think we should think more about the solution

Reply:
Is that why the recursive_search call is separated from the main method now, I maybe need to add an empty emit_block after that to set the caller as fully processed, but the solution looks fine to me in this case